src/graph/sampler.cc

```diff
@@ -111,6 +113,257 @@ class ArrayHeap {
+  std::vector<DType> _prob;     // categorical distribution
+  std::vector<DType> cdf;       // cumulative distribution function
+  std::vector<bool> used;       // indicate availability, activated when replace=false;
+  std::vector<Idx> id_mapping;  // indicate index mapping, activated when replace=false;
+
+  inline Idx map(Idx x) const {
+    if (replace)
+      return x;
+    else
+      return id_mapping[x];
+  }
+
+  void rebuild(const std::vector<DType>& prob) {
+    N = 0;
+    accum = 0.;
+    taken = 0.;
+    if (!replace)
+      id_mapping.clear();
+    cdf.clear();
+    cdf.push_back(0);
+    for (Idx i = 0; i < prob.size(); ++i)
+      if (!used[i]) {
+        ++N;
+        accum += prob[i];
+        if (!replace)
+          id_mapping.push_back(i);
+        cdf.push_back(accum);
+      }
+    if (N == 0) LOG(FATAL) << "Cannot take more sample than population when 'replace=false'";
+  }
+
+ public:
+  void reinit_state(const std::vector<DType>& prob) {
+    used.resize(prob.size());
+    if (!replace)
+      _prob = prob;
+    std::fill(used.begin(), used.end(), false);
+    rebuild(prob);
+  }
+
+  explicit CDFSampler(const std::vector<DType>& prob) {
+    reinit_state(prob);
+  }
+
+  ~CDFSampler() {}
+
+  inline Idx draw(RandomEngine *re) {
+    if (!replace) {
+      if (2 * taken >= accum)
+        rebuild(_prob);
+      while (true) {
+        DType p = re->Uniform<DType>(0., accum);
+        Idx rst = map(std::lower_bound(cdf.begin(), cdf.end(), p) - cdf.begin() - 1);
+        DType cap = _prob[rst];
+        if (!used[rst]) {
+          used[rst] = true;
+          taken += cap;
+          return rst;
+        }
+      }
+    }
+    DType p = re->Uniform<DType>(0., accum);
+    return map(std::lower_bound(cdf.begin(), cdf.end(), p) - cdf.begin() - 1);
+  }
+};
+
+/*
+ * TreeSampler is used to sample elements from a given discrete categorical distribution.
+ * Algorithm: create a heap that stores accumulated likelihood of its leaf descendents.
+ * Reference: https://blog.smola.org/post/1016514759
+ * Sampler building complexity: O(n)
+ * Sample w/ and w/o replacement complexity: O(log n)
+ */
+template <
+    typename Idx,
+    typename DType,
+    bool replace>
+class TreeSampler {
+ private:
+  std::vector<DType> weight;      // accumulated
+  Idx N, num_leafs;
+
+ public:
+  void reinit_state(const std::vector<DType>& prob) {
+    std::fill(weight.begin(), weight.end(), 0);
+    for (int i = 0; i < prob.size(); ++i)
+      weight[num_leafs + i] = prob[i];
+    for (int i = num_leafs - 1; i >= 1; --i)
+      weight[i] = weight[i * 2] + weight[i * 2 + 1];
+  }
+
+  explicit TreeSampler(const std::vector<DType>& prob) {
+    num_leafs = 1;
+    while (num_leafs < prob.size())
+      num_leafs *= 2;
+    N = num_leafs * 2;
+    weight.resize(N);
+    reinit_state(prob);
+  }
+
+  inline Idx draw(RandomEngine *re) {
+    Idx cur = 1;
+    while (cur < num_leafs) {
+      DType p = re->Uniform<DType>(0., weight[cur]);
```

Comment:
The original ArrayHeap only roll the dice once.

Reply:
DONE.